API update, build fixes, add some API flows

m-ildefons opened the pull request:
- API update:
Add a v1.3 API reference source and a source for the current development version.

- Build fixes:
Upgrade the openapi-docs plugin for Docusaurus to a stable version
Downgrade Docusaurus to 2.4.0, since 2.4.1 is incompatible with the openapi-docs plugin. This allows us to generate API reference docs from swagger.json files. These swagger.json files can be generated from the Golang type definitions of our CRDs in the harvester/harvester repository
The readme is also updated to add instructions how to maintain the API docs.

- Remove generated content
Existing generated content may confuse the openapi-docs plugin for docusaurus. To make sure that a clean git repository will also always be a clean basis for modifying the documentation, including the API reference docs, the generated files are removed. They can easily be re-generated if needed.

- Add some API flows.
Add documented API flows for uploading a VM image, creating a VM and registering a Harvester cluster in a Rancher Manager instance.
There are still many flows missing, but these three should serve as good examples, how our documentation could include both UI based flows and API based flows in the future:

https://github.com/harvester/docs/assets/17141774/2e56501c-cf9e-4eb5-bdad-235fdbf3c391

jillian-maroket commented on docs/upload-image.md:
```diff
@@ -20,15 +20,43 @@ Currently, there are three ways that are supported to create an image: uploading
+</TabItem>
+<TabItem value="api" label="API">
+
+To import a virtual machine image from a URL via the API, create a
+`VirtualMachineImage` object
```
```suggestion
To import a virtual machine image from a repository using the API, create a `VirtualMachineImage` object. You must specify a URL that can be accessed from the cluster.

Example:
```

jillian-maroket commented on docs/upload-image.md:
````diff
@@ -20,15 +20,43 @@ Currently, there are three ways that are supported to create an image: uploading
+  checksum: 80c27afb7cd791ac86ee1b0b0c572a242f6142579db5beac841e71151d370cd6
+```
+
+Refer to [the reference docs](./api/create-namespaced-virtual-machine-image)
+for the full specifications.
````
```suggestion
For more information, see the [API reference](./api/create-namespaced-virtual-machine-image).
```

jillian-maroket commented on docs/vm/create-vm.md:
```diff
@@ -43,6 +46,51 @@ Please refer to [this page](./create-windows-vm.md) for creating Windows virtual
+</TabItem>
+<TabItem value="api" label="API">
+
+You can create virtual machines through the K8s API by creating a
+`VirtualMachine` object.
```
```suggestion
To create virtual machines using the Kubernetes API, create a `VirtualMachine` object.
```

jillian-maroket commented on docs/vm/create-vm.md:
````diff
@@ -43,6 +46,51 @@ Please refer to [this page](./create-windows-vm.md) for creating Windows virtual
+            pod: {}
+```
+
+Check the [API reference](../api/create-namespaced-virtual-machine) for full specifications.
````
```suggestion
For more information, see the [API reference](../api/create-namespaced-virtual-machine).
```

jillian-maroket reviewed (commented):
@m-ildefons I reviewed the doc updates but not the readme. I'll go through it once I have a clearer idea of how we should proceed with the API content. 

As for Docusaurus, I've been informed by the doc team leads that we must upgrade to v3 in the near future (to align with the other ECM projects). 3.0.0-beta.x (beta) of the openapi-docs plugin is supposedly compatible with Docusaurus 3.0.1-3.1.1 as of last month. I'll provide more details after I sync up with the doc team members who have successfully upgraded their implementations.

m-ildefons commented:
Thanks @jillian-maroket ,
I'd like to avoid rolling the docusaurus 3 migration into _this_ PR as well, since it's gotten uncomfortably large already.
I've applied your suggestions, which also fixed the automatic numbering of bullet points in `docs/rancher/virtualization-management.md`

The API content has changed only in two ways:
1) I added the swagger.json files for the current master branch and v1.3 branch of Harvester. This should automatically reflect any changes that happened to the API on a technical level.
2) I removed the generated content and generated content only. Keeping it under version control isn't necessarily helpful, since we can easily generate it again from the sources, which still are under version control in the repo.
Keeping the generated content around has a specific problem, where the docusaurus openapi plugin sometimes doesn't like to update the files, if they already exist.
Since these generated files are never directly edited by a human, I don't see a reason to keep them in the repo, much like we don't keep Harvester's compiled binaries in _it's_ repo.

jillian-maroket reviewed (approved):
LGTM. Thanks for making all these changes. I wasn't thinking of adding the DS upgrade to this PR's scope. Upgrading to v3 seems to require considerable prep work.
